### PR TITLE
[WEB-3071] smart-on-fhir launch endpoint

### PR DESCRIPTION
**Pull request by krystophv**
[WEB-3071] initial work for smart-on-fhir launch endpoint

### CODE REVIEW

File: app/keycloak.js [hunk 1]
```diff
@@ -19,26 +19,47 @@ export const setTokenRefresh = (keycloak) => {
...
 
 export const updateKeycloakConfig = (info, store) => {
   if (!(isEmpty(info) || isEqual(_keycloakConfig, info))) {
+    const urlParams = new URLSearchParams(window.location.search);
+    const iss = sessionStorage.getItem('smart_iss') || urlParams.get('iss');
+    const launch = sessionStorage.getItem('smart_launch') || urlParams.get('launch');
+
+    let clientConfig;
+
+    if (iss && launch) {
```
Comment (henry-tp-public) [inline app/keycloak.js:34]: **Security Concern: Session Storage Risk**

Using `sessionStorage` for sensitive FHIR parameters without encryption poses a security risk. Consider encrypting these sensitive parameters or using more secure storage mechanisms. At minimum, add validation to ensure stored values haven't been tampered with.

```javascript
// Consider adding validation/encryption
const iss = validateAndDecrypt(sessionStorage.getItem('smart_iss')) || urlParams.get('iss');
```

File: app/pages/login/login.js [hunk 2]
```diff
@@ -94,6 +96,27 @@ export let Login = withTranslation()(class extends React.Component {
       redirectUri += dest;
     }
 
+    const urlParams = new URLSearchParams(routerState?.location?.search);
```
Comment (henry-tp-public) [inline app/pages/login/login.js:99]: **Security: URL Parameter Validation Needed**

Direct URL parameter extraction and storage without validation could be vulnerable to XSS or injection attacks. Add input validation for URL parameters (iss, launch, correlation_id).

```javascript
// Add validation helpers
const validateFhirIss = (iss) => {
  // Validate iss format and whitelist known domains
  return /^https:\/\/[\w.-]+\/fhir/.test(iss) ? iss : null;
};

const iss = validateFhirIss(urlParams.get('iss'));
const launch = urlParams.get('launch')?.replace(/[^a-zA-Z0-9-_]/, '') || null;
```

File: app/pages/patients/patients.js [hunk 3]
```diff
@@ -448,6 +449,7 @@ export function mapStateToProps(state) {
     selectedClinicId: state.blip.selectedClinicId,
     showingWelcomeMessage: state.blip.showingWelcomeMessage,
     user,
+    isSmartOnFhirMode: state.blip.smartCorrelationId !== null,
```
Comment (henry-tp-public) [inline app/pages/patients/patients.js]: **Code Consistency: Use Existing Selector**

Using direct state property access with magic strings is inconsistent with the rest of the codebase. Use the existing `selectIsSmartOnFhirMode(state)` selector for consistency.

```javascript
// Replace this line:
// isSmartOnFhirMode: state.blip.smartCorrelationId !== null,
// With:
isSmartOnFhirMode: selectIsSmartOnFhirMode(state),
```

File: app/components/chart/basics.js [hunk 4]
```diff
@@ -46,6 +46,7 @@ class Basics extends Component {
     updateChartPrefs: PropTypes.func.isRequired,
     uploadUrl: PropTypes.string.isRequired,
     removeGeneratedPDFS: PropTypes.func.isRequired,
+    isSmartOnFhirMode: PropTypes.bool.isRequired,
```
Comment (henry-tp-public) [inline app/components/chart/basics.js:54]: **PropTypes Consistency Issue**

Inconsistent PropTypes declarations across chart components. Some mark `isSmartOnFhirMode` as `isRequired` while others don't for the same prop. Consider standardizing this - if the prop is always passed down, it should consistently be marked as `isRequired`.

Also appears in: `app/components/chart/daily.js`, `app/components/chart/trends.js`, etc.

File: app/pages/login/login.js [hunk 5]
```diff
@@ -94,6 +96,27 @@ export let Login = withTranslation()(class extends React.Component {
...
+    if (iss && launch && !correlationId) {
+      correlationId = crypto.randomUUID();
+
+      sessionStorage.setItem('smart_correlation_id', correlationId);
+      sessionStorage.setItem('smart_iss', iss);
+      sessionStorage.setItem('smart_launch', launch);
+
+      this.props.setSmartCorrelationId(correlationId);
+    }
+
```
Comment (henry-tp-public) [inline app/pages/login/login.js:119]: **Code Complexity: Extract Method**

The render method has grown quite complex with multiple nested conditions for Smart-on-FHIR handling. Consider extracting the Smart-on-FHIR URL parameter logic into a separate method like `handleSmartOnFhirParams()` for better readability and maintainability.

File: app/core/selectors.js [hunk 6]
```diff
@@ -65,75 +65,97 @@ export const selectPatientSharedAccounts = createSelector([state => state.blip],
...
-};
-
-export const selectPermsOfLoggedInUser = (state) => {
-  let permsOfLoggedInUser = null;
-
-  if (state.blip.allUsersMap) {
-    if (state.blip.currentPatientInViewId) {
-      permsOfLoggedInUser = state.blip.selectedClinicId
+    return clinicPatient;
+  }
```
Comment (henry-tp-public) [inline app/core/selectors.js:102]: **Good Practice: Selector Performance**

✅ Excellent use of `createSelector` for the new `selectIsSmartOnFhirMode` selector. This provides proper memoization and follows the established patterns in the codebase.

File: app/pages/smartonfhir/smartonfhir.js [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+        }
+      }
+
+      const patientInfo = smartOnFhirData.patients?.[correlationId];
+      if (!patientInfo) {
+        setError(ErrorMessages.ERR_SMARTONFHIR_PATIENT_INFO_NOT_FOUND);
+        setIsProcessing(false);
+        return;
+      }
+
```
Comment (henry-tp-public) [inline app/pages/smartonfhir/smartonfhir.js]: **Error Handling Enhancement**

The error handling covers multiple states but could benefit from retry logic for network failures and clearer user guidance for recoverable errors. Consider adding:

1. Retry mechanism for API failures
2. More specific error messages for different failure types
3. Recovery actions where possible

File: app/components/footer/footer.js [hunk 8]
```diff
@@ -1,91 +1,96 @@
...
+import _ from 'lodash';
+import Version from '../version';
+import { Flex } from 'theme-ui';
+import { useSelector } from 'react-redux';
+import { selectIsSmartOnFhirMode } from '../../core/selectors';
+
+const t = i18next.t.bind(i18next);
+
+const Footer = ({ version, location, trackMetric }) => {
+  const inSmartOnFhirMode = useSelector(state => selectIsSmartOnFhirMode(state));
```
Comment (henry-tp-public) [inline app/components/footer/footer.js:13]: **Component Responsibility Concern**

The Footer component now depends on Redux state management, which breaks component purity. Consider passing `isSmartOnFhirMode` as a prop from the parent component instead of using `useSelector` directly to maintain better separation of concerns.

File: app/keycloak.js [hunk 9]
```diff
@@ -134,12 +160,11 @@ export const keycloakMiddleware = (api) => (storeAPI) => (next) => (action) => {
...
   const store = useStore();
-  let Wrapper = React.Fragment;
-  let wrapperProps = props;
-  if (keycloakConfig?.url) {
-    Wrapper = ReactKeycloakProvider;
-    wrapperProps = {
-      ...wrapperProps,
+
+  const { Wrapper, wrapperProps } = React.useMemo(() => {
+    const Wrapper = keycloakConfig?.url ? ReactKeycloakProvider : React.Fragment;
```
Comment (henry-tp-public) [inline app/keycloak.js:165]: **Performance Optimization**

✅ Good use of `useMemo` for expensive operations. The component re-rendering logic is well-optimized and follows React best practices.

File: app/pages/smartonfhir/smartonfhir.js [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+  const navigateTo = useCallback((path) =>
+    dispatch(push(path)),
+    [dispatch]
+  );
+
+  const [isProcessing, setIsProcessing] = useState(false);
+  const [error, setError] = useState(null);
+
+  // Hide Zendesk widget in Smart-on-FHIR mode
+  useEffect(() => {
```
Comment (henry-tp-public) [inline app/pages/smartonfhir/smartonfhir.js:71]: **Code Organization: Extract Utility**

The Zendesk hiding logic with retry mechanism could be extracted into a utility function for reusability across the application, especially if other components need to interact with Zendesk in Smart-on-FHIR mode.

```javascript
// utils/zendesk.js
export const hideZendeskWidget = (maxAttempts = 50) => {
  let attempts = 0;
  let timeoutId;
  
  const checkZendesk = () => {
    if (window.zE) {
      window.zE('webWidget', 'hide');
      return;
    }
    
    attempts++;
    if (attempts < maxAttempts) {
      timeoutId = setTimeout(checkZendesk, 100);
    }
  };
  
  checkZendesk();
  return () => timeoutId && clearTimeout(timeoutId);
};
```

**Review by henry-tp-public — COMMENTED** [review]
## Code Review Summary

This is a comprehensive and well-structured implementation of Smart-on-FHIR integration. The code follows established patterns and demonstrates good understanding of the existing codebase architecture.

### Key Strengths:
- ✅ Proper use of Redux selectors and state management patterns
- ✅ Good performance optimizations with `useMemo` and `createSelector`
- ✅ Comprehensive error handling with user-friendly messages
- ✅ Consistent prop threading throughout the component hierarchy
- ✅ Well-documented API methods

### Main Areas for Improvement:
1. **Security**: Add validation/encryption for sensitive parameters in sessionStorage
2. **Code Consistency**: Use existing selectors instead of direct state access
3. **Component Responsibility**: Reduce direct Redux dependencies in presentational components
4. **PropTypes Consistency**: Standardize required/optional prop declarations

### Security Priority:
The session storage of sensitive FHIR parameters should be addressed before merging, as this could pose security risks in production environments.

Overall, this is solid work that integrates well with the existing codebase. The suggested improvements are primarily around security hardening and maintaining consistency with established patterns.

File: app/components/clinic/WorkspaceSwitcher.js [hunk 11]
```diff
@@ -27,7 +27,7 @@ import Popover from '../elements/Popover';
 import { colors } from '../../themes/baseTheme';
 
 export const WorkspaceSwitcher = props => {
-  const { t, api, trackMetric } = props;
+  const { t, api, trackMetric, isSmartOnFhirMode } = props;
```
Comment (clintonium-119) [inline app/components/clinic/WorkspaceSwitcher.js]: You're passing `isSmartOnFhirMode` in as a prop here, whereas all other functional components seem to be using the custom selector.  Should this be changed to use the selector for consistency?

File: app/components/navbar/DesktopNavbar/DesktopNavbar.js [hunk 12]
```diff
@@ -127,7 +128,11 @@ export default withTranslation()(class extends React.Component {
   renderMiddleSection = () => {
     if (personUtils.isClinicianAccount(this.props.user)) {
       return (
-        <WorkspaceSwitcher api={this.props.api} trackMetric={this.props.trackMetric} />
+        <WorkspaceSwitcher
+          api={this.props.api}
+          trackMetric={this.props.trackMetric}
+          isSmartOnFhirMode={this.props.isSmartOnFhirMode}
```
Comment (clintonium-119) [inline app/components/navbar/DesktopNavbar/DesktopNavbar.js]: If you decide to use the selector approach in `WorkspaceSwitcher`, the `isSmartOnFhirMode` prop can be removed here.

File: app/pages/patients/patients.js (hunk 3 above)
Comment (clintonium-119) [reply app/pages/patients/patients.js]: This seems like a reasonable selection to me.  Any reason to not use the selector here, and instead opt for a null check?

File: app/pages/smartonfhir/NoPatientMatch.js [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+        </Box>
+        <Box as="li" sx={{  }}>
+          <Body1>
+            {t('Enter the required information, making sure the MRN and birthdate match the EHR')}
+          </Body1>
+        </Box>
+      </Box>
+      <Box sx={{ mb: 2 }}>
+          <Body1>
+            {t('After this, you’ll be able to open their Tidepool account from your EHR system.')}
```
Comment (clintonium-119) [inline app/pages/smartonfhir/NoPatientMatch.js]:
```suggestion
            {t('After this, you\'ll be able to open their Tidepool account from your EHR system.')}
```
Just converting to a basic single quote.

File: app/pages/smartonfhir/smartonfhir.js [hunk 14]
```diff
@@ -0,0 +1,208 @@
...
+              return;
+            }
+            if (results.length > 1) {
+              handleError(ErrorMessages.ERR_SMARTONFHIR_MULTIPLE_PATIENTS_FOUND, 'Direct Connect Multiple Patients Found');
+              return;
+            }
+
+            const patient = results[0]?.patient;
+            if (!patient || !patient.id) {
+              handleError('Invalid patient data received', 'Direct Connect Patient Lookup Failure');
```
Comment (clintonium-119) [inline app/pages/smartonfhir/smartonfhir.js]: Perhaps this should be added to ErrorMessages like the others.

File: app/pages/smartonfhir/smartonfhir.js [hunk 15]
```diff
@@ -0,0 +1,208 @@
...
+          });
+       });
+     }
+   }, [smartOnFhirData, isProcessing, working.fetchingPatients.inProgress, api, fetchPatients, navigateTo, error, smartCorrelationId, setSmartCorrelationId, windowObj, trackMetric, loggedInUserId, getClinics, handleError]);
+
+  if (isProcessing || working.fetchingPatients.inProgress) {
+    return (
+      <SmartOnFhirLayout>
+        <Box sx={{ textAlign: 'center' }}>
+          <Body1>{ErrorMessages.ERR_SMARTONFHIR_LOADING_PATIENT_DATA}</Body1>
```
Comment (clintonium-119) [inline app/pages/smartonfhir/smartonfhir.js]: This seems like it doesn't belong in ErrorMessages

File: app/pages/smartonfhir/smartonfhir.js [hunk 16]
```diff
@@ -0,0 +1,208 @@
...
+          )}
+        </Box>
+      </SmartOnFhirLayout>
+    );
+  }
+
+  return (
+    <SmartOnFhirLayout>
+      <Box sx={{ textAlign: 'center' }}>
+        <Body1>{ErrorMessages.ERR_SMARTONFHIR_INITIALIZING}</Body1>
```
Comment (clintonium-119) [inline app/pages/smartonfhir/smartonfhir.js]: Again, not sure if it makes sense to have this in the ErrorMessages export

File: test/unit/components/navbar/MobileNavbar/Menu.test.js [hunk 17]
```diff
@@ -73,6 +74,10 @@ describe('MobileNavbar/Menu', () => {
     handleViewAccountSettings.resetHistory();
     handleLogout.resetHistory();
 
+    if (wrapper && wrapper.unmount && wrapper.length > 0) {
```
Comment (clintonium-119) [inline test/unit/components/navbar/MobileNavbar/Menu.test.js]: Just curious why there are so many of these unmount checks/calls. Were they failing if only performed in one spot?

File: test/unit/redux/actions/async.test.js [hunk 18]
```diff
@@ -9493,5 +9600,162 @@ describe('Actions', () => {
...
+        expect(actions[1].error.message).to.equal(ErrorMessages.ERR_FETCHING_PATIENTS);
+        expect(actions[1].meta.apiError).to.equal(err);
+
+        expect(api.patient.getAll.calledOnce).to.be.true;
+        expect(api.patient.getAll.calledWith(options)).to.be.true;
+        expect(callback.calledOnce).to.be.true;
+        expect(callback.calledWith(err)).to.be.true;
+      });
+
+      it('should use a noop function if no callback is provided', () => {
```
Comment (clintonium-119) [inline test/unit/redux/actions/async.test.js]: Is this test really doing anything?  Can it be removed?

File: app/components/clinic/WorkspaceSwitcher.js (hunk 11 above)
Comment (krystophv) [reply app/components/clinic/WorkspaceSwitcher.js]: Yeah, I think this one was earlier on and I was having issues determining which components to go with the selector vs. prop drilling (and issues around just how many unit tests I'd have to update to have a provider if I added the selector). But, as you point out - this one's already functional. Updated. 👍🏼

File: test/unit/components/navbar/MobileNavbar/Menu.test.js (hunk 17 above)
Comment (krystophv) [reply test/unit/components/navbar/MobileNavbar/Menu.test.js]: I think this leaked from another local branch where I was starting to play around with React 19 updates where there's some _interesting_ behavior around mount/unmount. I've removed the extraneous ones.

File: test/unit/redux/actions/async.test.js (hunk 18 above)
Comment (krystophv) [reply test/unit/redux/actions/async.test.js]: Yeah, this one was a suggestion from an AI to cover the edge case where a callback isn't provided, just because it's technically possible. In practice, it'll never get called without a callback, so we can drop it. 👍🏼

File: app/pages/patients/patients.js (hunk 3 above)
Comment (krystophv) [reply app/pages/patients/patients.js]: Ah yeah, this is an artifact of having a custom hook at one point instead of the dedicated selector. Updated.
